source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2488,9 +2056,8 @@ void COpenGLDriver::SAuxContext::flushState(GL_STATE_BITS stateBits)
         {
             COpenGLExtensionHandler::extGlBindBuffersRange(GL_UNIFORM_BUFFER, 0u, MAX_UBO_BINDING_COUNT, ubonames, ubooffsets, ubosizes);
```

Comment:
why are you binding the full ranges!?

Only bind the range you need (at descriptor set granularity OFC), you've went from one extreme to the other.

Reply:
do i have to unbind previous bindings?

Reply:
elaborate?

Reply:
example:
we have bound desc set which occupies SSBO binding points 0,1,2
another desc set is going to be bound at the same index which occupies just SSBO binding point 0. Do i have to unbind SSBOs from binding points 1 and 2?

Reply:
yes you do, but it still comes out at 5 multibinds top.

This is why, descriptor sets N, N+1, ... will be disturbed by pipeline compatibility rules if the cumulative total of OpenGL bindpoint (UBO, SSBO) bindings will be different for descriptor set N .

Hence first A>=0 decriptor sets will be undisturbed, then B>=0 descriptor sets starting at index A will be disturbed (replaced by null descriptor set), then C>=0 descriptor sets starting at index A+B will be replaced, then D>=0 descriptor sets starting at index A+B+c will be disturbed.

A particular set can only stay wholly bound or wholly disturbed.
`A` needs to equal a *contiguous* subset starting at 0 of "identically defined" sets common in previous and current pipeline. So range A and B each must encompass whole descriptor sets in the current pipeline layout (not the old), C naturally encompasses whole descriptor sets by definition, whereas D encompasses the difference between the end of the old total binding range and new total binding range. Note that D>0 if and only if current pipeline layout has more descriptors of a particular type than the old pipeline layout.

Is it clear?

You just need the previous pipeline layout to figure out the indices and the ranges.

Reply:
@Crisspl this is still an issue in the latest commit. To recap only the following are possible:

# Everything changes

Then you do 1 multibind per resource type X (UBO,SSBO,Texture) with `count = previous pipeline layout total count of resource of type X across all descriptor sets`, since you're unbinding all array/pointer parameters to OpenGL multibind functions can be `nullptr`

# N (>0) descriptor sets stay the same (after validation rules)

There are two stages to this, bind new and unbind old.

## Bind New 

Nullptr (no descriptor) counts as a "new" descriptor binding.

Then you do 1 multibind per resource type X per new descriptor set bound, with `first = previous pipeline layout total count of resource of type X across descriptor sets in slots [0,N)`  the pointer/array parameters shall be sourced **directly** from precomputed (if desc. set nullptr, then parameters are nullptr) arrays inside the new descriptor sets in slots [N, N + New Descriptor Set Count] 

## Unbind Old

For every resource type X, you take the difference between `previous pipeline layout total count of resource of type X across all descriptor sets` and `new pipeline layout total count of resource of type X across all descriptor sets`

If the difference is positive (previous layout had more bindings, then you do 1 multibind per resource type X (UBO,SSBO,Texture) with `first = previous pipeline layout total count of resource of type X across all descriptor sets` and `count = the above mentioned difference`, since you're unbinding all array/pointer parameters to OpenGL multibind functions can be `nullptr`

Reply:
Had slight Vulkan spec clarification, written down in #376